File: tom_som_rust_runtime/src/yaml.rs
//! `yaml` — a minimal, dependency-free YAML reader for the constrained
//! `*.docspecs.yaml` subset the codec emits, a faithful port of the Go
//! `yaml.go`.
//!
//! The runtime ships with no external libraries, so this hand-rolled
//! indentation parser stands in for a YAML crate. It supports exactly what the
//! encoder produces (and what the self-verify round-trip probe needs): block
//! mappings, block sequences (`- item`), literal block scalars (`|`, `|-`, `|2`,
//! `|2-` …), JSON-quoted flow scalars, plain integers, and the empty flow
//! collections `{}` / `[]`.
//!
//! The dynamic value model is [`YamlValue`]: mappings are maps, sequences are
//! vectors, scalars are strings or integers. JSON-quoted scalars are parsed via
//! the hand-rolled [`crate::json`] parser.

use std::collections::BTreeMap;

use crate::json::Json;

/// A parsed YAML value.
#[derive(Debug, Clone, PartialEq)]
pub enum YamlValue {
    Map(BTreeMap<String, YamlValue>),
    Seq(Vec<YamlValue>),
    Str(String),
    Int(i64),
}

impl YamlValue {
    fn empty_map() -> YamlValue {
        YamlValue::Map(BTreeMap::new())
    }

    /// Returns the value at `key` when this is a map, else `None`.
    pub fn get(&self, key: &str) -> Option<&YamlValue> {
        match self {
            YamlValue::Map(m) => m.get(key),
            _ => None,
        }
    }

    /// Returns the map entries when this is a map, else `None`.
    pub fn as_map(&self) -> Option<&BTreeMap<String, YamlValue>> {
        match self {
            YamlValue::Map(m) => Some(m),
            _ => None,
        }
    }

    /// Returns the elements when this is a sequence, else `None`.
    pub fn as_seq(&self) -> Option<&[YamlValue]> {
        match self {
            YamlValue::Seq(s) => Some(s),
            _ => None,
        }
    }

    /// Returns the string when this is a string scalar, else `None`.
    pub fn as_str(&self) -> Option<&str> {
        match self {
            YamlValue::Str(s) => Some(s),
            _ => None,
        }
    }

    /// Returns the integer when this is an integer scalar, else `None`.
    pub fn as_int(&self) -> Option<i64> {
        match self {
            YamlValue::Int(n) => Some(*n),
            _ => None,
        }
    }

    /// Returns the scalar rendered as a string (string verbatim, integer
    /// formatted), or `""` for non-scalars.
    pub fn scalar_string(&self) -> String {
        match self {
            YamlValue::Str(s) => s.clone(),
            YamlValue::Int(n) => n.to_string(),
            _ => String::new(),
        }
    }
}

fn yaml_leading(line: &str) -> usize {
    line.bytes().take_while(|&b| b == b' ').count()
}

fn yaml_is_integer(s: &str) -> bool {
    if s.is_empty() {
        return false;
    }
    let bytes = s.as_bytes();
    let mut i = 0;
    if bytes[0] == b'-' || bytes[0] == b'+' {
        if bytes.len() == 1 {
            return false;
        }
        i = 1;
    }
    bytes[i..].iter().all(|b| b.is_ascii_digit())
}

fn yaml_find_colon(content: &str) -> Option<usize> {
    let bytes = content.as_bytes();
    if content.starts_with('"') {
        let mut i = 1;
        while i < bytes.len() {
            let c = bytes[i];
            if c == b'\\' {
                i += 2;
                continue;
            }
            if c == b'"' {
                i += 1;
                break;
            }
            i += 1;
        }
        return content[i..].find(':').map(|idx| i + idx);
    }
    content.find(':')
}

fn yaml_parse_scalar_key(key_part: &str) -> String {
    if key_part.starts_with('"') {
        if let Ok(Json::Str(v)) = Json::parse(key_part) {
            return v;
        }
        return String::new();
    }
    key_part.to_string()
}

struct YamlReader {
    lines: Vec<String>,
    idx: usize,
}

impl YamlReader {
    fn next_significant(&self, from: usize) -> Option<usize> {
        for i in from..self.lines.len() {
            let t = self.lines[i].trim();
            if t.is_empty() || t.starts_with('#') {
                continue;
            }
            return Some(i);
        }
        None
    }

    fn parse_node(&mut self, min_indent: usize) -> YamlValue {
        let i = match self.next_significant(self.idx) {
            Some(i) => i,
            None => return YamlValue::empty_map(),
        };
        let line = &self.lines[i];
        let indent = yaml_leading(line);
        if indent < min_indent {
            return YamlValue::empty_map();
        }
        let content = &line[indent..];
        if content == "-" || content.starts_with("- ") {
            self.parse_sequence(indent)
        } else {
            self.parse_mapping(indent)
        }
    }

    fn parse_mapping(&mut self, indent: usize) -> YamlValue {
        let mut m: BTreeMap<String, YamlValue> = BTreeMap::new();
        loop {
            let i = match self.next_significant(self.idx) {
                Some(i) => i,
                None => break,
            };
            let line = self.lines[i].clone();
            let cur_indent = yaml_leading(&line);
            if cur_indent != indent {
                break; // dedent → belongs to the parent.
            }
            self.idx = i + 1;
            let content = &line[indent..];
            let colon = match yaml_find_colon(content) {
                Some(c) => c,
                None => continue, // not a mapping entry — tolerate and skip.
            };
            let key = yaml_parse_scalar_key(content[..colon].trim());
            let rest = content[colon + 1..].trim().to_string();
            let value = if rest.is_empty() {
                self.parse_nested_after_key(indent)
            } else if rest.starts_with('|') || rest.starts_with('>') {
                YamlValue::Str(self.parse_block_scalar(&rest, indent))
            } else {
                self.parse_flow_scalar(&rest)
            };
            m.insert(key, value);
        }
        YamlValue::Map(m)
    }

    fn parse_sequence(&mut self, indent: usize) -> YamlValue {
        let mut list = Vec::new();
        loop {
            let i = match self.next_significant(self.idx) {
                Some(i) => i,
                None => break,
            };
            let line = self.lines[i].clone();
            let cur_indent = yaml_leading(&line);
            if cur_indent != indent {
                break;
            }
            let content = &line[indent..];
            if content != "-" && !content.starts_with("- ") {
                break;
            }
            self.idx = i + 1;
            let item_text = if content == "-" {
                String::new()
            } else {
                content[2..].trim().to_string()
            };
            if item_text.is_empty() {
                list.push(self.parse_nested_after_key(indent));
            } else if item_text.starts_with('|') || item_text.starts_with('>') {
                list.push(YamlValue::Str(self.parse_block_scalar(&item_text, indent)));
            } else {
                list.push(self.parse_flow_scalar(&item_text));
            }
        }
        YamlValue::Seq(list)
    }

    fn parse_nested_after_key(&mut self, parent_indent: usize) -> YamlValue {
        let i = match self.next_significant(self.idx) {
            Some(i) => i,
            None => return YamlValue::empty_map(),
        };
        let child_indent = yaml_leading(&self.lines[i]);
        if child_indent <= parent_indent {
            return YamlValue::empty_map();
        }
        self.parse_node(child_indent)
    }

    fn parse_flow_scalar(&self, s: &str) -> YamlValue {
        let t = s.trim();
        if t == "{}" {
            return YamlValue::empty_map();
        }
        if t == "[]" {
            return YamlValue::Seq(Vec::new());
        }
        if t.starts_with('"') {
            if let Ok(Json::Str(v)) = Json::parse(t) {
                return YamlValue::Str(v);
            }
            return YamlValue::Str(t.to_string());
        }
        if yaml_is_integer(t) {
            if let Ok(n) = t.parse::<i64>() {
                return YamlValue::Int(n);
            }
        }
        YamlValue::Str(t.to_string())
    }

    /// Parses a literal block scalar starting at `header` (e.g. `|2-`) whose body
    /// lines follow. `key_indent` is the indentation of the owning key/sequence
    /// line.
    fn parse_block_scalar(&mut self, header: &str, key_indent: usize) -> String {
        let hbytes = header.as_bytes();
        let mut p = 1; // skip the leading '|' (or '>').
        let digits_start = p;
        while p < hbytes.len() && hbytes[p].is_ascii_digit() {
            p += 1;
        }
        let explicit_indent: Option<usize> = if p > digits_start {
            header[digits_start..p].parse::<usize>().ok()
        } else {
            None
        };
        let mut chomp = b' ';
        while p < hbytes.len() {
            let c = hbytes[p];
            if c == b'-' || c == b'+' {
                chomp = c;
            }
            p += 1;
        }

        // Collect the raw body: blank lines plus any line indented past the key.
        let mut raw_body: Vec<String> = Vec::new();
        while self.idx < self.lines.len() {
            let l = self.lines[self.idx].clone();
            if l.trim().is_empty() {
                raw_body.push(String::new());
                self.idx += 1;
                continue;
            }
            if yaml_leading(&l) <= key_indent {
                break;
            }
            raw_body.push(l);
            self.idx += 1;
        }
        // Drop trailing blank lines that belong to the following node.
        while raw_body.last().map(|l| l.is_empty()).unwrap_or(false) {
            raw_body.pop();
        }

        let content_indent = match explicit_indent {
            Some(ei) => key_indent + ei,
            None => {
                let mut min = usize::MAX;
                for l in &raw_body {
                    if !l.is_empty() {
                        let ld = yaml_leading(l);
                        if ld < min {
                            min = ld;
                        }
                    }
                }
                if min == usize::MAX {
                    key_indent + 1
                } else {
                    min
                }
            }
        };

        let mut body: Vec<String> = Vec::new();
        for l in &raw_body {
            if l.is_empty() {
                body.push(String::new());
            } else {
                let strip = content_indent.min(l.len());
                body.push(l[strip..].to_string());
            }
        }

        let joined = body.join("\n");
        match chomp {
            b'-' => joined.trim_end_matches('\n').to_string(),
            b'+' => joined,
            _ => {
                // Clip (default): exactly one trailing newline when non-empty.
                let clipped = joined.trim_end_matches('\n');
                if clipped.is_empty() {
                    String::new()
                } else {
                    format!("{}\n", clipped)
                }
            }
        }
    }
}

/// Parses `text` into the value model (a map or sequence).
pub fn yaml_parse(text: &str) -> YamlValue {
    let lines: Vec<String> = text.split('\n').map(|s| s.to_string()).collect();
    let mut y = YamlReader { lines, idx: 0 };
    if y.next_significant(0).is_none() {
        return YamlValue::empty_map();
    }
    y.parse_node(0)
}

/// Parses `text`, returning a map (empty when not a mapping).
pub fn yaml_parse_map(text: &str) -> YamlValue {
    match yaml_parse(text) {
        v @ YamlValue::Map(_) => v,
        _ => YamlValue::empty_map(),
    }
}

File: tom_som_go_runtime/yaml.go
package somruntime

// yaml.go — a minimal, dependency-free YAML reader for the constrained
// `*.docspecs.yaml` subset the SpecDocumentYaml codec emits, a faithful port of
// `tom_som_dart_runtime/lib/src/yaml.dart` (and the TypeScript `yaml.ts`).
//
// The runtime ships with no external libraries, so this hand-rolled indentation
// parser stands in for package:yaml / PyYAML. It supports exactly what the
// encoder produces (and what the self-verify round-trip probe needs): block
// mappings, block sequences ("- item"), literal block scalars (|, |-, |2, |2- …),
// JSON-quoted flow scalars, plain integers, and the empty flow collections
// {} / [].
//
// The dynamic value model is interface{}: mappings are *YamlMap (insertion
// ordered — Go's native maps are unordered, but the hierarchical codec walks
// keys in document order, exactly like the JS/TS objects and Python dicts of
// the other ports), sequences are []interface{}, scalars are string or int.
// JSON-quoted scalars are parsed with encoding/json.

import (
	"encoding/json"
	"strings"
)

// YamlMap is an insertion-ordered string-keyed mapping — the Go analogue of a
// JavaScript object / Python dict for the parsed YAML value model. Iterating
// Keys() yields keys in first-insertion order (a later Set of an existing key
// replaces the value but keeps the original position), which is what makes the
// hierarchical decode walk deterministic and byte-compatible with the other
// language ports.
type YamlMap struct {
	keys   []string
	values map[string]interface{}
}

// NewYamlMap returns an empty ordered mapping.
func NewYamlMap() *YamlMap {
	return &YamlMap{values: map[string]interface{}{}}
}

// Set stores value under key, preserving the key's first-insertion position.
func (m *YamlMap) Set(key string, value interface{}) {
	if _, ok := m.values[key]; !ok {
		m.keys = append(m.keys, key)
	}
	m.values[key] = value
}

// Get returns the value stored under key, and ok=false when absent.
func (m *YamlMap) Get(key string) (interface{}, bool) {
	v, ok := m.values[key]
	return v, ok
}

// GetOr returns the value stored under key, or nil when absent.
func (m *YamlMap) GetOr(key string) interface{} {
	return m.values[key]
}

// Has reports whether key is present.
func (m *YamlMap) Has(key string) bool {
	_, ok := m.values[key]
	return ok
}

// Keys returns the keys in insertion order (the backing slice; callers must
// not mutate it).
func (m *YamlMap) Keys() []string {
	return m.keys
}

// Len returns the number of entries.
func (m *YamlMap) Len() int {
	return len(m.keys)
}

func yamlNewMap() *YamlMap {
	return NewYamlMap()
}

func yamlLeading(line string) int {
	n := 0
	for n < len(line) && line[n] == ' ' {
		n++
	}
	return n
}

func yamlIsInteger(s string) bool {
	if len(s) == 0 {
		return false
	}
	i := 0
	if s[0] == '-' || s[0] == '+' {
		i = 1
		if len(s) == 1 {
			return false
		}
	}
	for ; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			return false
		}
	}
	return true
}

func yamlFindColon(content string) int {
	if strings.HasPrefix(content, "\"") {
		i := 1
		for i < len(content) {
			c := content[i]
			if c == '\\' {
				i += 2
				continue
			}
			if c == '"' {
				i++
				break
			}
			i++
		}
		idx := strings.Index(content[i:], ":")
		if idx < 0 {
			return -1
		}
		return i + idx
	}
	return strings.Index(content, ":")
}

func yamlParseScalarKey(keyPart string) string {
	if strings.HasPrefix(keyPart, "\"") {
		var v string
		if err := json.Unmarshal([]byte(keyPart), &v); err != nil {
			return ""
		}
		return v
	}
	return keyPart
}

type yamlReader struct {
	lines []string
	idx   int
}

func (y *yamlReader) nextSignificant(from int) int {
	for i := from; i < len(y.lines); i++ {
		t := strings.TrimSpace(y.lines[i])
		if t == "" || strings.HasPrefix(t, "#") {
			continue
		}
		return i
	}
	return -1
}

func (y *yamlReader) parseNode(minIndent int) interface{} {
	i := y.nextSignificant(y.idx)
	if i == -1 {
		return yamlNewMap()
	}
	line := y.lines[i]
	indent := yamlLeading(line)
	if indent < minIndent {
		return yamlNewMap()
	}
	content := line[indent:]
	if content == "-" || strings.HasPrefix(content, "- ") {
		return y.parseSequence(indent)
	}
	return y.parseMapping(indent)
}

func (y *yamlReader) parseMapping(indent int) interface{} {
	m := yamlNewMap()
	for {
		i := y.nextSignificant(y.idx)
		if i == -1 {
			break
		}
		line := y.lines[i]
		curIndent := yamlLeading(line)
		if curIndent != indent {
			break // dedent → belongs to the parent; over-indent shouldn't occur.
		}
		y.idx = i + 1
		content := line[indent:]
		colon := yamlFindColon(content)
		if colon < 0 {
			continue // not a mapping entry — tolerate and skip.
		}
		key := yamlParseScalarKey(strings.TrimSpace(content[:colon]))
		rest := strings.TrimSpace(content[colon+1:])
		var value interface{}
		if rest == "" {
			value = y.parseNestedAfterKey(indent)
		} else if strings.HasPrefix(rest, "|") || strings.HasPrefix(rest, ">") {
			value = y.parseBlockScalar(rest, indent)
		} else {
			value = y.parseFlowScalar(rest)
		}
		m.Set(key, value)
	}
	return m
}

func (y *yamlReader) parseSequence(indent int) interface{} {
	list := []interface{}{}
	for {
		i := y.nextSignificant(y.idx)
		if i == -1 {
			break
		}
		line := y.lines[i]
		curIndent := yamlLeading(line)
		if curIndent != indent {
			break
		}
		content := line[indent:]
		if content != "-" && !strings.HasPrefix(content, "- ") {
			break
		}
		y.idx = i + 1
		itemText := ""
		if content != "-" {
			itemText = strings.TrimSpace(content[2:])
		}
		if itemText == "" {
			list = append(list, y.parseNestedAfterKey(indent))
		} else if strings.HasPrefix(itemText, "|") || strings.HasPrefix(itemText, ">") {
			list = append(list, y.parseBlockScalar(itemText, indent))
		} else {
			list = append(list, y.parseFlowScalar(itemText))
		}
	}
	return list
}

func (y *yamlReader) parseNestedAfterKey(parentIndent int) interface{} {
	i := y.nextSignificant(y.idx)
	if i == -1 {
		return yamlNewMap()
	}
	childIndent := yamlLeading(y.lines[i])
	if childIndent <= parentIndent {
		return yamlNewMap()
	}
	return y.parseNode(childIndent)
}

func (y *yamlReader) parseFlowScalar(s string) interface{} {
	t := strings.TrimSpace(s)
	if t == "{}" {
		return yamlNewMap()
	}
	if t == "[]" {
		return []interface{}{}
	}
	if strings.HasPrefix(t, "\"") {
		var v string
		if err := json.Unmarshal([]byte(t), &v); err == nil {
			return v
		}
		return t
	}
	if yamlIsInteger(t) {
		return atoi(t)
	}
	return t
}

// parseBlockScalar parses a literal block scalar starting at header (e.g. "|2-")
// whose body lines follow. keyIndent is the indentation of the owning
// key/sequence line.
func (y *yamlReader) parseBlockScalar(header string, keyIndent int) string {
	p := 1 // skip the leading '|' (or '>').
	explicitIndent := -1
	digitsStart := p
	for p < len(header) && header[p] >= '0' && header[p] <= '9' {
		p++
	}
	if p > digitsStart {
		explicitIndent = atoi(header[digitsStart:p])
	}
	chomp := byte(' ')
	for p < len(header) {
		c := header[p]
		if c == '-' || c == '+' {
			chomp = c
		}
		p++
	}

	// Collect the raw body: blank lines plus any line indented past the key.
	var rawBody []string
	for y.idx < len(y.lines) {
		l := y.lines[y.idx]
		if strings.TrimSpace(l) == "" {
			rawBody = append(rawBody, "")
			y.idx++
			continue
		}
		if yamlLeading(l) <= keyIndent {
			break
		}
		rawBody = append(rawBody, l)
		y.idx++
	}
	// Drop trailing blank lines that actually belong to the following node.
	for len(rawBody) > 0 && rawBody[len(rawBody)-1] == "" {
		rawBody = rawBody[:len(rawBody)-1]
	}

	var contentIndent int
	if explicitIndent >= 0 {
		contentIndent = keyIndent + explicitIndent
	} else {
		contentIndent = int(^uint(0) >> 1) // max int
		for _, l := range rawBody {
			if l != "" {
				if ld := yamlLeading(l); ld < contentIndent {
					contentIndent = ld
				}
			}
		}
		if contentIndent == int(^uint(0)>>1) {
			contentIndent = keyIndent + 1
		}
	}

	var body []string
	for _, l := range rawBody {
		if l == "" {
			body = append(body, "")
		} else {
			strip := contentIndent
			if strip > len(l) {
				strip = len(l)
			}
			body = append(body, l[strip:])
		}
	}

	joined := strings.Join(body, "\n")
	if chomp == '-' {
		return strings.TrimRight(joined, "\n")
	}
	if chomp == '+' {
		return joined
	}
	// Clip (default): exactly one trailing newline when non-empty.
	clipped := strings.TrimRight(joined, "\n")
	if clipped == "" {
		return ""
	}
	return clipped + "\n"
}

// YamlParse parses text into the value model (a *YamlMap or []interface{}).
func YamlParse(text string) interface{} {
	lines := strings.Split(text, "\n")
	y := &yamlReader{lines: lines, idx: 0}
	first := y.nextSignificant(0)
	if first == -1 {
		return yamlNewMap()
	}
	return y.parseNode(0)
}

// YamlParseMap parses text, returning a mapping (empty when not a mapping).
func YamlParseMap(text string) *YamlMap {
	v := YamlParse(text)
	if m, ok := v.(*YamlMap); ok {
		return m
	}
	return yamlNewMap()
}
